### PR TITLE
DOC-2732: Updated details on hash-sharded indexes

### PR DESCRIPTION
**Pull request by nickvigilante**
Completes the following issues:

- DOC-2732
- DOC-2244
- DOC-2218
- DOC-1693
- DOC-874
- DOC-811
- DOC-2572
- DOC-2552
- DOC-2353

### CODE REVIEW

**Comment by nickvigilante** [comment]
Related to https://github.com/cockroachdb/cockroach/pull/79082.

**Review by chengxiong-ruan — COMMENTED** [review]
good job on these writing! LGTM mostly, just two nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13250#-:-Mzuz6aF9Nzme_RazhJQ:b-wsgyjm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @devadvocado and @nickvigilante)

___
*[v22.1/hash-sharded-indexes.md, line 36 at r1](https://reviewable.io/reviews/cockroachdb/docs/13250#-Mzuwm_3CKyAZq3IEkSY:-Mzuwm_3CKyAZq3IEkSZ:b-jtlpjf) ([raw file](https://github.com/cockroachdb/docs/blob/48ea9e1d5a4d4c12ecabfd6602cff0ee18aeb060/v22.1/hash-sharded-indexes.md#L36)):*
> an inverse proportion to the performance of reading sequential data or ranges of data
@ajwerner I'm not sure this is true, my understanding is that all shards are queried in parallel and a merge sort on the read node. But there's definitely penalty on read.
___
*[v22.1/hash-sharded-indexes.md, line 105 at r1](https://reviewable.io/reviews/cockroachdb/docs/13250#-MzuyMC-7FNxJMUK9AfX:-MzuyMC-7FNxJMUK9AfY:b-rqtx3l) ([raw file](https://github.com/cockroachdb/docs/blob/48ea9e1d5a4d4c12ecabfd6602cff0ee18aeb060/v22.1/hash-sharded-indexes.md#L105)):*
> ```Markdown
> ### Create a hash-sharded secondary index with a different `bucket_count`
> 
> You can specify a different `bucket_count` on a hash-sharded index to optimize either write performance or sequential read performance on a table:
> ```

worth mentioning that `bucket_count` is one of the supported storage parameter :)
https://www.cockroachlabs.com/docs/stable/sql-grammar.html#storage_parameter
It was only used for inverted index (see in current doc: https://www.cockroachlabs.com/docs/stable/create-index.html)
We make the bucket count as a storage param to be more postgresql compatible...

**Review by nickvigilante — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13250)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @devadvocado, and @nickvigilante)

___
*[v22.1/hash-sharded-indexes.md, line 105 at r1](https://reviewable.io/reviews/cockroachdb/docs/13250#-MzuyMC-7FNxJMUK9AfX:-Mzv0eFF61zZa25xgeWi:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/48ea9e1d5a4d4c12ecabfd6602cff0ee18aeb060/v22.1/hash-sharded-indexes.md#L105)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

worth mentioning that `bucket_count` is one of the supported storage parameter :)
https://www.cockroachlabs.com/docs/stable/sql-grammar.html#storage_parameter
It was only used for inverted index (see in current doc: https://www.cockroachlabs.com/docs/stable/create-index.html)
We make the bucket count as a storage param to be more postgresql compatible...
</blockquote></details>

Done.

**Comment by chengxiong-ruan** [comment]
:lgtm: mod the statement about performance. need another around from @ajwerner

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 3 of 13 files at r1, 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13250#-:-N-PPoUg5gJxyOVlCtCl:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @devadvocado)

___
*[v22.1/hash-sharded-indexes.md, line 36 at r1](https://reviewable.io/reviews/cockroachdb/docs/13250#-Mzuwm_3CKyAZq3IEkSY:-N-POvza536YmIxK6ABt:b2p7ws) ([raw file](https://github.com/cockroachdb/docs/blob/48ea9e1d5a4d4c12ecabfd6602cff0ee18aeb060/v22.1/hash-sharded-indexes.md#L36)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

> an inverse proportion to the performance of reading sequential data or ranges of data
@ajwerner I'm not sure this is true, my understanding is that all shards are queried in parallel and a merge sort on the read node. But there's definitely penalty on read.
</blockquote></details>

I agree that the language is a little bit opaque. Maybe use more plain english. How about something like:

A larger number of buckets allows for greater load-balancing and thus greater write throughput. More buckets disadvantages operations which need to scan over the data to fulfill their query; such queries will now need to scan over each bucket and combine the results.

**Comment by nickvigilante** [comment]
@ajwerner @chengxiong-ruan Done. Sending off for writer peer review.

@gemma-shay Can you take a look at this?

**Review by gemma-shay — APPROVED** [review]
lgtm!

File: _includes/v22.1/performance/alter-primary-key-hash-sharded.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    serial_number VARCHAR,
+    event_id UUID,
+    ts TIMESTAMP,
+    data JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts) USING HASH
+);
+~~~
+
+You can change an existing primary key to use hash sharding simply by adding the `USING HASH` clause at the end of the key definition:
```
[the remark this answers is not in the document]
Comment (nickvigilante) [reply _includes/v22.1/performance/alter-primary-key-hash-sharded.md:17]: Ah, I should have caught that. Thank you!

**Pull request merged**